piet-svg/src/lib.rs

```diff
@@ -0,0 +1,475 @@
+    ImageFormat, InterpolationMode, IntoBrush, LineCap, LineJoin, StrokeStyle,
+};
+use svg::node::Node;
+
+type Result<T> = std::result::Result<T, Error>;
+
+pub struct RenderContext {
+    stack: Vec<State>,
+    state: State,
+    doc: svg::Document,
+    next_id: u64,
+    text: Text,
+}
+
+impl RenderContext {
+    pub fn new() -> Self {
+        Self {
+            stack: Vec::new(),
+            state: State::default(),
+            doc: svg::Document::new(),
+            next_id: 0,
+            text: Text,
+        }
+    }
+
+    pub fn write(&self, writer: impl io::Write) -> io::Result<()> {
+        svg::write(writer, &self.doc)
+    }
+
+    fn new_id(&mut self) -> Id {
+        let x = Id(self.next_id);
+        self.next_id += 1;
+        x
+    }
+}
+
+impl piet::RenderContext for RenderContext {
+    type Brush = Brush;
+
+    type Text = Text;
+    type TextLayout = TextLayout;
+
+    type Image = Image;
+
+    fn status(&mut self) -> Result<()> {
+        Ok(())
+    }
+
+    fn clear(&mut self, color: Color) {
+        let brush = color.make_brush(self, || Rect::ZERO);
+        let mut rect = svg::node::element::Rectangle::new()
+            .set("width", "100%")
+            .set("height", "100%")
+            .set("fill", brush.val());
+        if let Some(id) = self.state.clip {
+            rect.assign("clip-path", format!("url(#{})", id.to_string()));
+        }
+        self.doc.append(rect);
+    }
+
+    fn solid_brush(&mut self, color: Color) -> Brush {
+        Brush {
+            kind: BrushKind::Solid(color),
+        }
+    }
+
+    fn gradient(&mut self, gradient: impl Into<FixedGradient>) -> Result<Brush> {
+        let id = self.new_id();
+        match gradient.into() {
+            FixedGradient::Linear(x) => {
+                let mut gradient = svg::node::element::LinearGradient::new()
+                    .set("gradientUnits", "userSpaceOnUse")
+                    .set("id", id)
+                    .set("x1", x.start.x)
+                    .set("y1", x.start.y)
+                    .set("x2", x.end.x)
+                    .set("y2", x.end.y);
+                for stop in x.stops {
+                    gradient.append(
+                        svg::node::element::Stop::new()
+                            .set("offset", stop.pos)
+                            .set("stop-color", fmt_color(&stop.color)),
+                    );
+                }
+                self.doc.append(gradient);
+            }
+            FixedGradient::Radial(x) => {
+                let mut gradient = svg::node::element::RadialGradient::new()
+                    .set("gradientUnits", "userSpaceOnUse")
+                    .set("id", id)
+                    .set("cx", x.center.x)
+                    .set("cy", x.center.y)
+                    .set("fx", x.center.x + x.origin_offset.x)
+                    .set("fy", x.center.y + x.origin_offset.y)
+                    .set("fr", 0)
```

Comment:
There are no `fr` in SVG.

Reply:
Yes, there is: https://svgwg.org/svg2-draft/pservers.html#RadialGradientElementFRAttribute

Oddly, MDN disagrees with the spec on the default value (50% vs 0% respectively).

Reply:
I see. I should clarify that I was talking about SVG 1.1. Which is a better target, since probably no one supports SVG 2 anyway.

SVG 2 is not finished yet, so everything is possible.

Reply:
Fair enough.